pkg/client/apiutil/dynamicrestmapper.go

```diff
@@ -29,34 +28,12 @@ import (
 	"k8s.io/client-go/restmapper"
 )
 
-// ErrRateLimited is returned by a RESTMapper method if the number of API
-// calls has exceeded a limit within a certain time period.
-type ErrRateLimited struct {
-	// Duration to wait until the next API call can be made.
-	Delay time.Duration
-}
-
-func (e ErrRateLimited) Error() string {
-	return "too many API calls to the RESTMapper within a timeframe"
-}
-
-// DelayIfRateLimited returns the delay time until the next API call is
-// allowed and true if err is of type ErrRateLimited. The zero
-// time.Duration value and false are returned if err is not a ErrRateLimited.
-func DelayIfRateLimited(err error) (time.Duration, bool) {
```

Comment:
We can also think about deprecating `DelayIfRateLimited` (and always returning `0, false`) instead of removing it.
This would make this PR 🐛 instead of ⚠️ FMPOV, so we could cherry-pick it to `release-0.x`.
WDYT?

Reply:
could also do a wrapper error that wraps both errors if it's a nokindmatcherror that's *caused* by rate limitting.  This has the downside of not working with `meta.IsNoMatchError` (but that function is actually not really correct in Go 1.13+) until they're updated to work with Go's 1.13 standard `errors` package, but has the advantage that it works with idiomatic Go:

```go
if errors.As(err, &meta.NoKindMatchError{}) {
  // no kind matched
}
```

as well as still allowing you to have `DelayIfRateLimited` work.

I'd imagine the implementation would look like:

```go
type ErrorNoMatch struct {
  Delay time.Duration
  matchError error
}

func (e *ErrNoMatch) Unwrap()  error {
  return e.matchError
}
// implement Error() as necessary

// alias for backwards compat

type ErrRateLimited = ErrNoMatch
```

WDYT?

Reply:
Thanks @DirectXMan12 for your suggestion. I like it wrt using go errors idiomatically and also keeping backwards compatability.

Although I am not voting totally against it, I want to note something:
TBH I don't see a real benefit of annotating the `NoKindMatchError` with a delay. I can't think of a use case where a controller/consumer would actually want to distinguish between a "normal" nokindmatch and a nokindmatch that might have been returned by a slightly outdated cache, and then wait until it can retry.
If you added a new CRD for example and you want to add a CR, then you will have to expect a slight delay until the resource comes available in the discovery API anyways, right?
Also, IISIC the code actually doesn't guarantee the caller, that they won't be rate limited again after waiting the returned delay (reservation is cancelled and other callers trying to rediscover might sneak in).

I would think of the specified rate limit as some kind of a SLA between the controller/consumer and the RESTMapper, which says "Please make sure, my discovery cache is not outdated by more than 1s" (if you specify a rate limiter with 1rps).
And if you're call was rate limited because some other rediscoveries exhausted the rate limited only a few milliseconds before yours, than your cache is very likely not to be outdated, although the rate limiter didn't allow the request.

It quite unfortunate, that upstream has not adapted go1.13 errors, yet. But I would expect that an implementation of the `meta.RESTMapper` interface is also working as expected with the provided helper functions in the `meta` package, like `meta.IsNoMatchError`.


As said, I'm not against implementing your approach, I just wanted to give some more background on the motivation why I started with the proposed implementation.